### ISSUE TITLE
Support the stem macros (math support)

### ISSUE THREAD

**Issue by jxxcarlson** [asciidoctor/asciidoctor-pdf#45]

**Comment by tstumm** [comment]
For LaTeX this could be done by handing the equation/formula to a LaTeX to SVG parser, whose output can easily be embedded. This shouldn't be that much work.

**Comment by mojavelinux** [comment]
The best LaTeX to SVG parser that I'm aware of is MathJax. Unfortunately, it only works in a browser. But, we have figured out a way to use it as a utility to generate an SVG by delegating to mathoid. See https://github.com/asciidoctor/asciidoctor-extensions-lab/tree/master/lib/mathoid-treeprocessor. I was planning on integrating this utility into Asciidoctor PDF.

If you know of a simpler utility, I'm all for exploring it. Ideally, we want to keep external dependencies and tools to an absolute minimum.

**Comment by mojavelinux** [comment]
The first step to solving this issue might just be to document how to use the aforementioned extension. I'm able to use it, but that doesn't mean anyone else can figure it out easily unless we document it :)

**Comment by tstumm** [comment]
How about using [Mathematical](https://github.com/gjtorikian/mathematical)?
Most of it's dependencies (pango, cairo) should be covered by GTK libs afaik.
It's a native ruby library which would make it easy to implement.

**Comment by mojavelinux** [comment]
Interesting. That looks like what GitHub is (or is planning) to use to support Math expressions on GitHub (given who is committing to it). Let's definitely consider it!

**Comment by tstumm** [comment]
I just build a library wrapper for it [here](https://github.com/tstumm/asciidoctor-lib-mathematical). I don't know if it's working, so we should run some tests with it.

**Comment by tstumm** [comment]
I've written a sample file and a test-case and it seems to be working pretty good.

**Comment by mojavelinux** [comment]
Cool! Super exciting!

I welcome you to send a pull request for your extension into the extensions lab (https://github.com/asciidoctor/asciidoctor-extensions-lab). That helps get it exposure and a stop towards official integration.

**Comment by jxxcarlson** [comment]
Fantastic!!

**Comment by tstumm** [comment]
@mojavelinux As you may have seen I've sent a pull request. No response yet, but I'm looking forward for a fast implemetation of this feature in asciidoctor-pdf (no more LaTeX for pdfs yeah!)

**Comment by jxxcarlson** [comment]
When we get this working, I will open a bottle!

**Comment by mojavelinux** [comment]
I apologize for the delay. This has been a crazy busy week. I'll be sure to
get to it by the weekend at the latest. If not, you can nudge me :)

Cheers,

-Dan

On Fri, Jun 26, 2015 at 6:08 AM, tstumm <notifications@github.com> wrote:

> @mojavelinux <https://github.com/mojavelinux> As you may have seen I've
> sent a pull request. No response yet, but I'm looking forward for a fast
> implemetation of this feature in asciidoctor-pdf (no more LaTeX for pdfs
> yeah!)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asciidoctor/asciidoctor-pdf/issues/45#issuecomment-115656859>
> .
>



-- 
Dan Allen | http://google.com/profiles/dan.j.allen

**Comment by jxxcarlson** [comment]
Yay!!!

**Comment by jxxcarlson** [comment]
Any chance that Asciidoctor-off will be able to handle Asciidoc-LaTeX files?  That would give us another rest selling point in the STEM. Communities.

**Comment by jxxcarlson** [comment]
Fantastic!!

**Comment by mojavelinux** [comment]
> Any chance that Asciidoctor-pdf will be able to handle Asciidoc-LaTeX files? 

In time, I'm sure of it as both mature and LaTeX and PDF have such a close relationship. For now, the focus is just to get Asciidoctor PDF out of alpha so we can start doing interesting things with it :)

**Comment by mojavelinux** [comment]
@tstumm This looks really nice. I'm quite impressed by mathematical and the extension that you wrote. 

It's important to note that right now, this will only handle stem blocks. We'll need to add another extension to handle inline. That's a bit trickier since it's not yet possible to override a built-in inline macro. We'll need to think about the best way to approach that, or perhaps we need a change to core to allow us to tap into the stem inline macro.

In order to use this extension or the mathoid one in Asciidoctor PDF, we'll need to graduate it from the extensions lab. I think a first step is just to document how someone can use the extension in the extensions lab so we get some feedback and testing. That way, no one is waiting for the graduation to happen. In the meantime, we can start to think about where we want to put it. Perhaps a gem named asciidoctor-mathematical.

**Comment by mojavelinux** [comment]
I forgot to mention that I sent revisions to your pull request. Once you look those over and merge or revise, I'll merge in the pull request.

**Comment by tstumm** [comment]
@mojavelinux Thank your for the commendation. I'm a ruby first-timer though.
I think changing the core would be the best solution, since this should be overridable anway.

Maybe we can reuse the readme as documentation. The code itself is rather trivial, so we should mainly document the needed dependencies (fonts etc.).

**Comment by mojavelinux** [comment]
Yes, for right now the README is the user manual. In the future, each component (e.g., asciidoctor-pdf) will have it's own user manual under the docs folder to keep the README can stay lean. However, not yet :)

@tstumm You did a nice job for a Ruby first timer. This is a great start and the rest will come quickly now that the ball is rolling. I think the biggest challenge for most users will be installing mathematical...because *a lot* of people have no development libraries installed. For me, I only had to install "flex" (whatever that is).

**Comment by tstumm** [comment]
Well, for me I had to install a bunch of libraries (because I hadn't looked up the dependencies, shame on me). On a debian or ubuntu system that should be fulfilled by one apt-get install command.
And on Windows... just kidding.

I think we can now mark this as "in progress"!

**Comment by mojavelinux** [comment]
Indeed!

I found the best way to discover dependencies is to use Docker. There's rarely anything in an official Docker image that wouldn't be available by default on a workstation.

...on Windows we can simply recommend using Cygwin (or Babun) which is actually a pretty good environment for normalizing the setup with that of Linux.

**Comment by tstumm** [comment]
Fortunately Lasem (Mathematical is just a wrapper for it) has a very detailed dependency list (including fonts. LaTeX equation look very good using Droid Sans. Not.).

Well, let's get that thing going!

**Comment by mojavelinux** [comment]
> For the runtime, the following ttf fonts should be installed: cmr10, cmmi10, cmex10 and cmsy10.
They are provided by the lyx-fonts package in fedora, and the ttf-lyx package in debian/ubuntu.

Ah! Now the equations look a heck of a lot better!

Here's a screenshot of lasem vs MathJax on my machine now that I have the fonts installed:

![math-sample](https://cloud.githubusercontent.com/assets/79351/8497219/c40b5f7c-2138-11e5-88cc-0391e94e385e.png)

**Comment by tstumm** [comment]
Is it just me, or does the Lasem font look a lot clearer?
Anyways, Lasem should be a lot faster than MathJax.

**Comment by mojavelinux** [comment]
Actually, I'm not sure installing the fonts made a difference. Either I already had them in another directory or it didn't pick them up. Either way, I get the same result.

> Lasem should be a lot faster than MathJax.

It's blistering fast in comparison. When the page loads, it makes MathJax look like it was late for work.

**Comment by mojavelinux** [comment]
I agree, the MathJax fonts are over styled (too much serif).

**Comment by tstumm** [comment]
Okay, try some integrals and sums without the font. You will see the difference.

**Comment by tstumm** [comment]
Could you try to adapt the extensions code in a local asciidoctor-pdf branch?
I already looked through your code, but I think you know better how the parser works.

I'd love to see an example pdf!

**Comment by mojavelinux** [comment]
It shouldn't be necessary to add anything to Asciidoctor PDF. The goal is that this just becomes another gem (for now, we have to build it locally, then install it). Once the gem is installed, you just add `-r mathematical-treeprocessor` (or something to that effect) and it gets used automatically. What we need to sort out is how to make this easy to do :)

**Comment by mojavelinux** [comment]
Basically, it will work just like Asciidoctor Diagram...which is an example of how it looks once it's all packaged and distributed.

**Comment by tstumm** [comment]
@mojavelinux Do you know how to build a gem package? I'd like to try it out and experiment a bit.

**Comment by mojavelinux** [comment]
There are several good resources on the topic.

* https://mattbrictson.com/building-a-gem-in-2015
* http://guides.rubygems.org/make-your-own-gem/
* http://www.smashingmagazine.com/2014/04/08/how-to-build-a-ruby-gem-with-bundler-test-driven-development-travis-ci-and-coveralls-oh-my/
* https://quickleft.com/blog/engineering-lunch-series-step-by-step-guide-to-building-your-first-ruby-gem/
* http://timelessrepo.com/making-ruby-gems
* https://blog.engineyard.com/2015/its-raining-gems
* https://howistart.org/posts/ruby/1

**Comment by tstumm** [comment]
So shall I merge the changes of yours to my repository and add a gem file, or shall I do a pull request again?

**Comment by tstumm** [comment]
I just edited my own repository: https://github.com/tstumm/mathematical-treeprocessor

The gem works, but the svg looks huge in pdfs - maybe scaling went wrong?

**Comment by tstumm** [comment]
![Scaling error](https://cloud.githubusercontent.com/assets/1056434/8518782/88c7256c-23cd-11e5-9759-6ea6b10ebd1b.png)

**Comment by mojavelinux** [comment]
> I just edited my own repository:

That's what I was going to suggest.

I do recommend that you rename to the gem to asciidoctor-mathematical or something similar. When you publish the gem, it goes into the global namespace of all gems, so it should have a clear association with the libraries it's integrating.

Once you do that, I'd definitely like to welcome the gem into the asciidoctor organization, if the community agrees.

> the svg looks huge in pdfs - maybe scaling went wrong?

You are waiting on the 1.5.0.alpha.9 release of Asciidoctor PDF which addresses this issue.

Until then, you can set the scaledwidth attribute to something other than 100% (which is the default). For instance:

```
attrs = { 'target' => image_target, 'alt' => alt_text, 'align' => 'center', 'scaledwidth' => '50%' }
```

Until 1.5.0.alpha.9, there's no way to easily set it to a native width as that was the bug that was fixed.

**Comment by mojavelinux** [comment]
Here's the related image scaling issue that was resolved in the upcoming release: https://github.com/asciidoctor/asciidoctor-pdf/issues/76

**Comment by tstumm** [comment]
Been there, done that https://github.com/tstumm/asciidoctor-mathematical

Well, I think waiting for the fix is the best solution.

**Comment by mojavelinux** [comment]
I should have alpha.9 out by the end of the week / weekend for sure.

**Comment by tstumm** [comment]
I checked out the current master branch and did some tests. SVGs now scale nicely but the size of the svg does not equal text size. Any ideas? I couldn't find any attributes for font size.

**Comment by mojavelinux** [comment]
> the size of the svg does not equal text size. 

I'm not clear what you mean by this. Do you mean the text size in the SVG is not equivalent to the text size in the document?

**Comment by tstumm** [comment]
That's exactly what I meant. The text size in the SVG is smaller than the text size in the document.

**Comment by mojavelinux** [comment]
The font size in the SVG and the font size in the document are independent things. That said, there may be some way we can influence the font size in the generated SVG to better align with the font size in the document.

**Comment by tstumm** [comment]
That was the response I was waiting for. Is there any method to get text size of the document in Treeprocessor?

**Comment by mojavelinux** [comment]
I could make it happen. I need to make a change, but very possible.

**Comment by ProgramFan** [comment]
Hello, I made some changes to asciidoctor-mathematical to implement high resolution PNG support as well as inline stem support. So inline stem (only latexmath, though) can be supported. The result with asciidoctor-pdf-1.5.0.alpha.9 looks promissing:
![result](https://cloud.githubusercontent.com/assets/32396/9109222/8be6e9f6-3c66-11e5-9612-776a058205a1.png)

I also created a pull request for my implementation to asciidoctor-mathematical. It would be nice to see this issue CLOSED. :smile:

**Comment by ProgramFan** [comment]
How about make this happen by v1.5.0.beta1? I am willing to help.

**Comment by paulvickers** [comment]
How did people get this to work?  I have a test document:
``` asciidoc
= Test book
Paul Vickers
:stem: latexmath

include::asciidoc/master.ad[]
```
And the include document:
``` asciidoc
== Hello World

How do you do
stem:[x\times y]
```
I run:
`asciidoctor-pdf -o test.pdf test.ad`
But the math output shows:
`x\times y`

What am I missing?

**Comment by jacksteeple** [comment]
@ProgramFan: can you post the example code for inline formulas, in my asciidoctor-pdf (V1.5.0) only the block element works (with low resolution), like this

[latexmath]
++++
k_{n+1} = n^2 + k_n^2 - k_{n-1}
++++

does not work: latexmath:[sqrt(4) = 2]

**Comment by ProgramFan** [comment]
@jacksteeple You have to use latex syntax for latexmath macro. Currently, my patch still lives in my own repository, namely [http://github.com/ProgramFan/asciidoctor-mathematical][]. So you have to use asciidoctor-mathematical from my repository to use inline latexmath.

**Comment by ProgramFan** [comment]
@paulvickers You need to install asciidoctor-mathematical and use it as an extension with the following command:

    asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf test.ad

**Comment by paulvickers** [comment]
@ProgramFan Thanks for the tip. I tried installing asciidoctor-mathematical but am not having success. I downloaded the zip archive and created a gem from the gemspec. I then tried a gem install but it wouldn't work. I see there a lots of dependencies, but it's not clear whether these should be installed using gem, or something else. Could you tell me exactly what you did to install asciidoctor-mathematical? Thanks. :-)

**Comment by paulvickers** [comment]
Update. I tried installing asciidoctor-mathematical yesterday, but that was from tstumm's repository and it didn't work. However, the one from your repository seems to install fine. I also did:
`gem install cairo`
and
`gem install pango`

But when I run 
`asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf sample.adoc`
I get:
`dyld: lazy symbol binding failed: Symbol not found: _pango_cairo_font_map_get_default
  Referenced from: /Library/Ruby/Gems/2.0.0/gems/mathematical-1.4.2/lib/mathematical/mathematical.bundle
  Expected in: flat namespace

dyld: Symbol not found: _pango_cairo_font_map_get_default
  Referenced from: /Library/Ruby/Gems/2.0.0/gems/mathematical-1.4.2/lib/mathematical/mathematical.bundle
  Expected in: flat namespace

Trace/BPT trap: 5'

**Comment by tstumm** [comment]
You need to install libcairo2 and libpango (via apt or yum). Maybe you also need -dev packages.

**Comment by paulvickers** [comment]
I'm running OS X 10.10.5 (Yosemite) so don't have apt or yum. Anyway, I had a look at the mathematical README and it seems there is a problem with Cairo 1.14.2 and Yosemite, so I have just uninstalled the Cairo gem and used Homebrew to install the earlier 1.12.16 version which is recommended by mathematical (gem install can't find v 1.12.16). But no luck. Still getting the errors above.

**Comment by tstumm** [comment]
Try
brew install pango

**Comment by jacksteeple** [comment]
@paulvickers: maybe this helps somebody: I used a Gemfile and the bundler (see http://bundler.io/gemfile.html) with:
===================
gem "asciidoctor"
gem "asciidoctor-pdf"
gem "asciidoctor-diagram"

gem "mathematical"
gem "asciidoctor-mathematical", :github => 'tstumm/asciidoctor-mathematical'
==================
und with the command "bundler" it downloads all the files with the dependencies

**Comment by jacksteeple** [comment]
@ProgramFan your changes where merged tonight to tstumm/asciidoctor-mathematical und works fine now in the pdf and in html :-) Thanx a lot!

**Comment by paulvickers** [comment]
Sigh... After much head scratching I think the reason it wasn't installing was due to a conflict between homebrew and macports. I removed that ambiguity and have now managed to install asciidoctor-mathematical on an OS X 10.9 system (avoiding the Yosemite problems highlighted yesterday). But, when I try compiling the sample file with:

    asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf sample.adoc
I now get the following error:

    /Users/paulvickers/Documents/Installation Sets/asciidoctor-mathematical-master/images/stem-d881ea909920c838a98ff43c5f1ffb41.png; image file is an unrecognised format
    asciidoctor: WARNING: could not embed image: /Users/paulvickers/Documents/Installation Sets/asciidoctor-mathematical-master/images/stem-7ad57f57f7601fadff003f5376571237.png; image file is an unrecognised format
    image file is an unrecognised format

Any ideas?

**Comment by jacksteeple** [comment]
my parenthesis has a wrong maybe font?
```
[latexmath]
++++
c = (a + b)^2 
++++
```
![klammer](https://cloud.githubusercontent.com/assets/13869689/9382909/ece06ac8-4744-11e5-8e52-3b3be5dc4c50.png)

Any ideas how to solve that?

**Comment by tstumm** [comment]
Install the fonts with:

cd ~/Library/Fonts
curl -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/cmex10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/cmmi10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/cmr10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/cmsy10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/esint10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/eufm10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/msam10.ttf \
     -LO http://mirrors.ctan.org/fonts/cm/ps-type1/bakoma/ttf/msbm10.ttf

**Comment by paulvickers** [comment]
Fonts were already installed, but still getting the `WARNING: could not embed image:` error

**Comment by jacksteeple** [comment]
@tstumm  thanx will try to install them on my linux system
@paulvickers : do you habe images in your adoc? does it work wihout them?

**Comment by paulvickers** [comment]
I'm using the sample document that came with asciidoctor-mathematical. The images are generated by the asciidoctor-mathematical when it runs. I have tried opening the generated pngs with Preview but they are, indeed, an invalid format.

**Comment by tstumm** [comment]
@jacksteeple Sorry those were the directions for OS X. Look at asciidoctor-mathematical readme. There is some information about font dependencies.

@paulvickers Can you give some information about the images? Maybe try to open them to validate if they are corrupt.

**Comment by jacksteeple** [comment]
@paulvickers did you installed the gems via a bundler or each manually? I use a bundler and here are my dependencies (see above my gem file):
GIT
  remote: git://github.com/asciidoctor/asciidoctor-pdf.git
  revision: 6675565a20828913785b8cc5e1d169ea09ea4d5e
  specs:
    asciidoctor-pdf (1.5.0.dev)
      asciidoctor (~> 1.5.0)
      prawn (>= 1.3.0, < 3.0.0)
      prawn-icon (= 0.6.4)
      prawn-svg (= 0.21.0)
      prawn-table (= 0.2.2)
      prawn-templates (= 0.0.3)
      safe_yaml (= 1.0.4)
      thread_safe (= 0.3.5)
      treetop (= 1.5.3)

GIT
  remote: git://github.com/tstumm/asciidoctor-mathematical.git
  revision: 99c7d29f653bdc90b4288d35cc3ebe1488baf81d
  specs:
    asciidoctor-mathematical (0.0.4)
      asciidoctor (~> 1.5, >= 1.5.0)
      mathematical (~> 1.4, >= 1.4.2)
      ruby-enum (~> 0.4)

GEM
  remote: https://rubygems.org/
  specs:
    Ascii85 (1.0.2)
    addressable (2.3.8)
    afm (0.2.2)
    asciidoctor (1.5.2)
    asciidoctor-diagram (1.3.0)
      asciidoctor (~> 1.5.0)
    coderay (1.1.0)
    css_parser (1.3.6)
      addressable
    hashery (2.1.1)
    i18n (0.7.0)
    mathematical (1.4.2)
      ruby-enum (~> 0.4)
    pdf-core (0.6.0)
    pdf-reader (1.3.3)
      Ascii85 (~> 1.0.0)
      afm (~> 0.2.0)
      hashery (~> 2.0)
      ruby-rc4
      ttfunk
    polyglot (0.3.5)
    prawn (2.0.2)
      pdf-core (~> 0.6.0)
      ttfunk (~> 1.4.0)
    prawn-icon (0.6.4)
      prawn (>= 1.1.0, < 3.0.0)
    prawn-svg (0.21.0)
      css_parser (~> 1.3)
      prawn (>= 0.8.4, < 3)
    prawn-table (0.2.2)
      prawn (>= 1.3.0, < 3.0.0)
    prawn-templates (0.0.3)
      pdf-reader (~> 1.3)
      prawn (>= 0.15.0)
    ruby-enum (0.4.0)
      i18n
    ruby-rc4 (0.1.5)
    safe_yaml (1.0.4)
    thread_safe (0.3.5)
    treetop (1.5.3)
      polyglot (~> 0.3)
    ttfunk (1.4.0)

PLATFORMS
  ruby

DEPENDENCIES
  asciidoctor
  asciidoctor-diagram
  asciidoctor-mathematical!
  asciidoctor-pdf!
  coderay
  mathematical

**Comment by ProgramFan** [comment]
@paulvickers I am sorry that I have no experience with Mac so I can not help. My dev platform is Fedora 22. Everything works when I install the lyx-fonts package.

**Comment by ProgramFan** [comment]
@jacksteeple I have not tried the parentheses yet (my fault). I will check it tomorrow when I have my computer at hand.

**Comment by paulvickers** [comment]
@jacksteeple thanks for the bundler idea. I'm new to Ruby, Gem, etc, so am not sure what to do here. What file do I need to create? You have a gemfile above with the following contents:

    gem "asciidoctor"
    gem "asciidoctor-pdf"
    gem "asciidoctor-diagram"
    
    gem "mathematical"
    gem "asciidoctor-mathematical", :github => 'tstumm/asciidoctor-mathematical'

Do I put this into a file called, say, `am.gemfile`? Then what? And where do I put that list of dependencies you listed in your recent post? Sorry to be a pain, but this whole thing is driving me slowly insane and I just want to be able to include math in pdf documents. :-)

**Comment by ProgramFan** [comment]
@jacksteeple You formulae works on my computer. Everything is correctly rendered. So what platform are you using? Please double check the font names to see if there are font name conflicts.

**Comment by jacksteeple** [comment]
@ProgramFan : thanx for checking, I will check the fonts. have to wait for our server guy. it is a debian system.

**Comment by jacksteeple** [comment]
@paulvickers if you have all the dependencies I have no idea, sorry :-( for the bundler here are the infos: http://bundler.io/ 
see there:
* install buindler (gem install bundler)
* write the Gemfile (you have it already)
* type: "bundle install"

**Comment by paulvickers** [comment]
In case this will mean anything to anyone, here's the result with the backtrace switched on:

```Failed to read PNG contents/Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/image_handler.rb:40:in `find': image file is an unrecognised format (Prawn::Errors::UnsupportedImageType)
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/images.rb:87:in `build_image_object'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/formatted_text/inline_image_arranger.rb:72:in `block in arrange_images'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/formatted_text/inline_image_arranger.rb:51:in `each'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/formatted_text/inline_image_arranger.rb:51:in `arrange_images'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/formatted_text/inline_image_arranger.rb:14:in `wrap'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text/formatted/box.rb:223:in `block (4 levels) in render'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/font.rb:101:in `font_size'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/prawn_ext/extensions.rb:237:in `font_size'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text/formatted/box.rb:216:in `block (3 levels) in render'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/pdf-core-0.6.0/lib/pdf/core/text.rb:182:in `text_rendering_mode'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text/formatted/box.rb:211:in `block (2 levels) in render'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/pdf-core-0.6.0/lib/pdf/core/text.rb:204:in `character_spacing'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text/formatted/box.rb:210:in `block in render'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/font.rb:207:in `save_font'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text/formatted/box.rb:209:in `render'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text.rb:376:in `fill_formatted_text_box'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text.rb:215:in `formatted_text'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/prawn-2.0.2/lib/prawn/text.rb:167:in `text'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:2211:in `typeset_text'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:1638:in `layout_prose'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:399:in `convert_paragraph'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:97:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:71:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `block in content'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `map'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `content'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:113:in `convert_content_for_block'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:320:in `convert_section'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:97:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:71:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `block in content'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `map'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:80:in `content'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/document.rb:1086:in `content'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:113:in `convert_content_for_block'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:164:in `convert_document'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/lib/asciidoctor-pdf/converter.rb:97:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/document.rb:1028:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor.rb:1499:in `convert'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/cli/invoker.rb:94:in `block in invoke!'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/cli/invoker.rb:86:in `each'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-1.5.2/lib/asciidoctor/cli/invoker.rb:86:in `invoke!'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/gems/asciidoctor-pdf-1.5.0.alpha.9/bin/asciidoctor-pdf:27:in `<top (required)>'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/bin/asciidoctor-pdf:23:in `load'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/bin/asciidoctor-pdf:23:in `<main>'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/bin/ruby_executable_hooks:15:in `eval'
	from /Users/paulvickers/.rvm/gems/ruby-2.1.2/bin/ruby_executable_hooks:15:in `<main>'```

**Comment by ProgramFan** [comment]
@paulvickers This backtrace only says the image file is broken. We have to look into why mathematical generates bad images. The first thing to do is to install mathematical and write a simple ruby script to test its png generation. I am planning to by a Macbook Pro so maybe I can look into this issue when I get it.

**Comment by ProgramFan** [comment]
@tstumm If there are compatibility problems with mathematical, we can use mathoid as an alternative.

**Comment by jacksteeple** [comment]
now it works :-)

the fonts were missing.

here is a link how to install fonts in linux (http://community.linuxmint.com/tutorial/view/29) and download the fonts described above in https://github.com/asciidoctor/asciidoctor-pdf/issues/45#issuecomment-132997163

**Comment by tstumm** [comment]
@jacksteeple Installing fonts on Fedora / Ubuntu is described in the readme of asciidoctor-mathematical.

**Comment by jacksteeple** [comment]
@tstumm: yes, thanx!!! now everything is clear. the problem is to find the problem ;-)

**Comment by jacksteeple** [comment]
@ProgramFan does the "less" - symbol work at your inline examples, I get an "unknown character":
```

works!
[latexmath]
+++++++
1 \leq 1
+++++++

works!
latexmath:[1 \leq 1]

works!
[latexmath]
+++++++
1 < 2
+++++++

does not work!
latexmath:[1 < 2]
```
![forms](https://cloud.githubusercontent.com/assets/13869689/9440373/e3ba9d24-4a6e-11e5-8e30-db70d96483da.jpg)

**Comment by ProgramFan** [comment]
@jacksteeple It does not work on my side, either. I will invesgitate into it.

**Comment by ProgramFan** [comment]
@jacksteeple @tstumm The 'Unknown character' is fixed.

**Comment by jxxcarlson** [comment]
Below is what I did to install mathematical and asciidoctor-mathematical on Mac OS:

```
git clone https://github.com/ProgramFan/asciidoctor-mathematical
cd asciidoctor-mathematical/
gem build asciidoctor-mathematical.gemspec 
gem install asciidoctor-mathematical-0.0.4.gem 
 brew update
 brew install cmake
 brew install glib
 brew install gdk
brew install gdk-pixbuf
brew install pango
gem install asciidoctor-mathematical-0.0.4.gem 
Building native extensions.  This could take a while...
Successfully installed mathematical-1.4.2
Successfully installed asciidoctor-mathematical-0.0.4
```

**Comment by jxxcarlson** [comment]
However, running

```
$ asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf sample.adoc
```

gives me the following error messages

```
Failed to read PNG contentsFailed to read PNG contentsFailed to read PNG contentsasciidoctor: WARNING: could not embed image: /Users/carlson/dev/git/asciidoctor-mathematical/images/stem-d881ea909920c838a98ff43c5f1ffb41.png; image file is an unrecognised format
asciidoctor: WARNING: could not embed image: /Users/carlson/dev/git/asciidoctor-mathematical/images/stem-7ad57f57f7601fadff003f5376571237.png; image file is an unrecognised format
image file is an unrecognised format
```

The files in 

```
/Users/carlson/dev/git/asciidoctor-mathematical/images
```

are damaged (are unreadable)

**Comment by paulvickers** [comment]
Thanks, @jxxcarlson. That's a relief that it's not just me then.

**Comment by tstumm** [comment]
Try adding
```
:mathematical-format: svg
```
to the top of your document. Maybe mathematical is failing at generating pngs.

**Comment by paulvickers** [comment]
Okay, just tried that. The contents of sample.adoc are now:

    = _Precompiled_ Math!
    :math:
    :mathematical-format: svg
    :imagesoutdir: images
    :stem: latexmath
    ## Equations!
   
    [stem]
    ++++
    k_{n+1} = n^2 + k_n^2 - k_{n-1}
    ++++
    
    Some useful text! Formula for quadratic root:
    
    [stem]
    ++++
    x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}
    ++++
    
    Inline equation works too! stem:[a^2+b^2=c^2]. Pretty nice, huh?

And the resultant error is now a single line:

    Failed to read PNG contentsimage file is an unrecognised format

**Comment by tstumm** [comment]
Okay this seems like a bug from asciidoctor-mathematical. Although I still don't know why png generation fails. I set up a Hackintosh and will try to investigate.

**Comment by ProgramFan** [comment]
@paulvickers Would you please try setting mathematical-format using command line? So that the inline macro processor can correctly use svg format:

    $ asciidoctor-pdf -r asciidoctor-mathematical -a mathematical-format=svg -o test.pdf sample.adoc

**Comment by paulvickers** [comment]
@ProgramFan Well, that doesn't generate any errors, but the pdf output has no math in it, and the svg files appear to be empty (though not malformatted).

**Comment by ProgramFan** [comment]
Nice try. Then we will wait for @tstumm 's result.

2015年8月28日星期五，Paul Vickers <notifications@github.com> 写道：

> @ProgramFan <https://github.com/ProgramFan> Well, that doesn't generate
> any errors, but the pdf output has no math in it, and the svg files appear
> to be empty (though not malformatted).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asciidoctor/asciidoctor-pdf/issues/45#issuecomment-135493792>
> .
>


-- 
Zhang YANG (杨章)

Institute of Applied Physics and Computational Mathematics.
Addr: No. 2 East Fenghao Rd., 100094, Beijing, China.
Tel: 86-10-59872424 (Office)
Email: zyangmath@gmail.com OR yang_zhang@iapcm.ac.cn

**Comment by jxxcarlson** [comment]
Running from the command line with

```
$ asciidoctor-pdf -r asciidoctor-mathematical -a mathematical-format=svg -o test.pdf sample.adoc
```
or inserting `:mathematical-format: svg` in the file
results in the production of `test.pdf` with no errors reported. The pdf 
document shows the text very nicely, but formulas are blank.  
I believe that they occupy the correct space.

Looking into  `asciidoctor-mathematical/images`, I found 
three `svg` files.  However,  the files "empty" -- have dimensions, 
but no content in the viewBox:

```
$ cat stem-4e4cad74f2dc2eb92cc5b8fbf76de692.svg 
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="60.700017pt" height="8.910156pt" viewBox="0 0 60.700017 8.910156" version="1.1">
</svg>
```

It is wonderful to see so much progress on the 
`mathematical` extension for  `asciidoctor-pdf`.
I look forward to a release and then later to making it work with 
[Asciidoctor-LaTeX](http://math.noteshare.io/titlepage/asciidoctor-latex-manual) --used as
the math-rendering engine for [Noteshare.io](http://math.noteshare.io).

**Comment by tstumm** [comment]
Well the issue seems to be reproducible (see https://github.com/gjtorikian/mathematical/issues/51).
I couldn't find a solution yet.

**Comment by nathany** [comment]
@tstumm There is a path issue when using the -D option to set a destination for the pdf files being generated. The SVG/PNG files are generated in the destination folder, but it is looking for those files alongside the *.adoc being processed.

asciidoctor: WARNING: image to embed not found or not readable: /Users/nathany/src/github.com/nathany/book/manuscript/stem-4818ca709a67edb34c070ab47627dc1c.svg

**Comment by nathany** [comment]
@jxxcarlson thanks for the installation instructions

note: you can put multiple packages on one line
```
brew install glib gdk-pixbuf cairo pango cmake gobject-introspection autoconf automake libtool
```

@tstumm This would be quite a bit easier if you were to release a gem to RubyGems, even if it's marked as a pre-release.

**Comment by nathany** [comment]
@ProgramFan The inline math is having a similar folder problem in my setup. It's putting files under my book/images folder (the imagedir I'm using for other images), but looking for them under book/manuscript/images.

asciidoctor: WARNING: image to embed not found or not readable: /Users/nathany/src/github.com/nathany/book/manuscript/images/stem-7964c6a339acf2ddea25a5ef0552b97e.png

**Comment by nathany** [comment]
@tstumm One more issue, which only occurs when requiring `-r asciidoctor-mathematical`.

tonnes of warnings like this.

```
asciidoctor: WARNING: dropping line containing reference to missing attribute: sourcedir
...
asciidoctor: WARNING: ch01.adoc: line 98: no callouts refer to list item 1
asciidoctor: WARNING: ch01.adoc: line 99: no callouts refer to list item 2
...
```

**Comment by tstumm** [comment]
@nathany Please open a separate issue at the asciidoctor-mathematical repo.

**Comment by tstumm** [comment]
Discussions related to OS X can be continued here: https://github.com/tstumm/asciidoctor-mathematical/issues/3

All other issues can be proclaimed at the issue page.

**Comment by tstumm** [comment]
By the way: asciidoctor-mathematical is now available on RubyGems!

**Comment by tstumm** [comment]
Since the main functionality is now given, I think we can close this issue now @mojavelinux.

**Comment by ProgramFan** [comment]
Thanks for all your effort. It's a nice job :+1: @tstumm

**Comment by nathany** [comment]
@tstumm For my purposes, I ended up using `&#8531;` for ⅓ rather than asciidoctor-mathematica, at least for now. Just so you know I haven't tested your latest changes when combined with the -D destination flag.

Thanks for releasing a gem, btw.

**Comment by JmyL** [comment]
`imagesoutdir` and `imagesdir` attribute works different in **asciidoctor-diagram** and **asciidoctor-mathematical**.

When I set `imagesoutdir` to `outimg` and `imagesdir` to 'img', asciidoctor-diagram generates `<img src="img/filename.png" ... >` but asciidoctor-mathematical generates `<img src="img/outimg/filename.png" ... >`.
And 'asciidoctor-diagram' puts it's output image to `./outimg` folder, and 'asciidoctor-mathematical' choose `./outimg` and `./images` folder as it's output destination.
**So both extensions are not working well.**
I can use asciidoctor-diagram with setting both attribute to same folder name like `images`, but asciidoctor-mathematical still isn't working because it references `./images/images/...`.

@mojavelinux Can you suggest a standard with these attributes?

**Comment by tstumm** [comment]
Well, `imagesdir` describes the main location of image files and `imagesoutdir` is for generated images. Since the images created by extensions are images used in the output file but maybe need to be saved separately we save it in the `imagesoutdir` which should be a subdirectory of `imagesdir` (makes sense).

ltdr; `img/outimage` should be perfectly right in your case. If you want it to save elsewhere just feed it a relational path (ex. `../teximg`) [dunno if this works, I guess it should]

**Comment by JmyL** [comment]
It makes sense. But generated images are saved in `outimage` folder, not `img/outimage`.
And` images` folder is also generated when using 'inline math'. It does not depended on the value of attributes.
Math block generates `<img src="img/outimage/...>` but Inline math generates `<img src="img/images/...>`.

**Comment by Dalar** [comment]
Hello,

I'm on OSX (have installed everything with `brew` and `gem` inside `rbenv`). 
```
$ asciidoctor-pdf -v
Asciidoctor PDF 1.5.0.alpha.11 using Asciidoctor 1.5.4 [http://asciidoctor.org]
Runtime Environment (ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin13]) (lc:UTF-8 fs:UTF-8 in:- ex:UTF-8)
```

When running `asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf sample.adoc` I get the following error:
```
/Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': incompatible library version - /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/mathematical-1.5.12/lib/mathematical/mathematical.bundle (fatal)
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/mathematical-1.5.12/lib/mathematical.rb:1:in `<top (required)>'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical/extension.rb:3:in `<top (required)>'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical.rb:1:in `require_relative'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical.rb:1:in `<top (required)>'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:133:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:133:in `rescue in require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:40:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-1.5.4/lib/asciidoctor/cli/options.rb:219:in `block in parse!'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-1.5.4/lib/asciidoctor/cli/options.rb:217:in `each'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-1.5.4/lib/asciidoctor/cli/options.rb:217:in `parse!'
	from /Users/amodig/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/asciidoctor-pdf-1.5.0.alpha.11/bin/asciidoctor-pdf:25:in `<top (required)>'
	from /Users/amodig/.rbenv/versions/2.3.0/bin/asciidoctor-pdf:23:in `load'
	from /Users/amodig/.rbenv/versions/2.3.0/bin/asciidoctor-pdf:23:in `<main>'
```

Installing `asciidoctor-mathematical` gave errors on creating documentation, but I turned that off. Otherwise successful. I don't know ruby so I don't know where to put my finger on...

**Comment by ProgramFan** [comment]
It looks something wrong happes to mathematical. Is it correctly installed?

**Comment by Dalar** [comment]
I followed the instructions [here](https://github.com/tstumm/asciidoctor-mathematical). Also checked the `mathematical` [dependencies](https://github.com/gjtorikian/mathematical/blob/master/script/bootstrap).

And I said wrong, `asciidoctor-mathematical` install is good. Just parsing the documentation for `mathematical` gives MIME errors like this:
```
Parsing documentation for mathematical-1.5.12
unknown encoding name "'application/x-tex'>" for ext/mathematical/mtex2MML/tests/fixtures/MathJax/LaTeXToMathML-out/above-below/accents-1.html, skipping
...
```
Otherwise the installation is "good".

I see that the `mathematical` library doesn't load, so perhaps I need to take this matter there. But if there's any other options (can I render pdf with `MathoidTreeprocessor`?) or advice, I'm glad to hear.

**Comment by igagis** [comment]
After installing `asciidoctor-mathematical` gem and trying to run it as

```
asciidoctor-pdf -r asciidoctor-diagram -r asciidoctor-mathematical mydoc.adoc
```

I get the following

```
/usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require': liblasem.so: cannot open shared object file: No such file or directory - /var/lib/gems/2.1.0/extensions/x86_64-linux/2.1.0/mathematical-1.5.12/mathematical/mathematical.so (LoadError)
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /var/lib/gems/2.1.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /var/lib/gems/2.1.0/gems/mathematical-1.5.12/lib/mathematical.rb:1:in `<top (required)>'
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /var/lib/gems/2.1.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /var/lib/gems/2.1.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical/extension.rb:3:in `<top (required)>'
	from /var/lib/gems/2.1.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical.rb:1:in `require_relative'
	from /var/lib/gems/2.1.0/gems/asciidoctor-mathematical-0.0.5/lib/asciidoctor-mathematical.rb:1:in `<top (required)>'
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:135:in `require'
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:135:in `rescue in require'
	from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:144:in `require'
	from /var/lib/gems/2.1.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	from /usr/lib/ruby/vendor_ruby/asciidoctor/cli/options.rb:220:in `block in parse!'
	from /usr/lib/ruby/vendor_ruby/asciidoctor/cli/options.rb:218:in `each'
	from /usr/lib/ruby/vendor_ruby/asciidoctor/cli/options.rb:218:in `parse!'
	from /var/lib/gems/2.1.0/gems/asciidoctor-pdf-1.5.0.alpha.11/bin/asciidoctor-pdf:25:in `<top (required)>'
	from /usr/local/bin/asciidoctor-pdf:23:in `load'
	from /usr/local/bin/asciidoctor-pdf:23:in `<main>'
```

**Comment by tstumm** [comment]
@igagis Seems like you could not install mathematical correctly. You can find further help on https://github.com/gjtorikian/mathematical

**Comment by igagis** [comment]
I had to manually copy liblasem.so from /var/lib/gems/2.1.0/gems/mathematical-1.5.12/ext/mathematical/lasem/build to /usr/lib

**Comment by mojavelinux** [comment]
Since there are multiple solutions for processing math (aka stem) equations, and those solutions require external dependencies, I think it's safe to say that this processing should be handled by an extension (at least for now).

One solution is asciidoctor-mathematical. The other is the soon to be asciidoctor-mathoid (which now resides in the [extensions-lab](https://github.com/asciidoctor/asciidoctor-extensions-lab) repository).

We still have some challenges to overcome, but I think we should address those challenges by submitting issues in the respective extension project.

Once we have determined that the extensions are both stable and reliable, we can consider enabling one of the extensions by default.

**Comment by mojavelinux** [comment]
I think it's really important that image generation be handled consistently between asciidoctor-diagram, asciidoctor-mathematical, and asciidoctor-mathoid. Since Asciidoctor Diagram is the most mature in this regard, it should be used as the reference.

**Comment by mojavelinux** [comment]
I do think it would be a good idea to add information to the README about asciidoctor-mathematical (and perhaps the mathoid extension too). @tstumm or @ProgramFan, would you be willing to submit a PR to the README that introduces asciidoctor-mathematical. I recommend adding a section named "Converting STEM Nodes" before the "Optional Scripts" section.

**Comment by mojavelinux** [comment]
I'll close the issue once asciidoctor-mathematical is mentioned in the README. I'll then add an introduction to the mathoid extension (though I'm not recommending it until it gets moved into its own repository and published to RubyGems.org).

**Comment by mojavelinux** [comment]
When we mention Asciidoctor Mathematical, we should also mention that it doesn't yet have support for table cells and refer to the issue.
